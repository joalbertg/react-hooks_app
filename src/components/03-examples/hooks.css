body {
  padding: 60px;
  background-color: black;
  color: white;
}

hr {
  border-top: 1px solid rgba(255, 255, 255, .5);
}

.wraper-buttons {
  display: flex;
  justify-content: space-between;
}

button {
  flex: 0 1 calc(50% - 1rem);
}

