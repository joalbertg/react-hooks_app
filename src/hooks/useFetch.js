import {useState, useEffect, useRef } from 'react';

export const useFetch = url => {
  const isMounted = useRef(true);
  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null
  });

  useEffect(() => {
    return () => {
      isMounted.current = false;
    }
  }, []);

  useEffect(() => {
    setState(currentState => ({ ...currentState, loading: true }));
      fetch(url)
        .then(resp => resp.json())
        .then(data => {

          //setTimeout(() => {

          // para setear el estado de manera segura
          if(isMounted.current) {
            setState(currentState => ({
              ...currentState,
              loading: false,
              data
            }));
          } else {
            console.log('setState not was called!!!');
          }

        //}, 2000);

        })
        .catch(() => {
          setState({
            data: null,
            loading: false,
            error: 'Request not found.'
          });
        });
  },[url]);

  return state;
}

