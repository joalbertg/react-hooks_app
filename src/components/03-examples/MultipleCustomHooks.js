import React from 'react';

import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';

import './hooks.css';

export const MultipleCustomHooks = () => {
  const { counter, increment, decrement, reset } = useCounter(1);

  if(counter <= 0) reset();

  const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
  // !data -> !undefined -> true
  // !true -> false
  const { author, quote } = !!data && data[0];

  return(
    <>
      <h1>BreakingBad Quotes</h1>
      <hr />

      { loading ?
          <div className="alert alert-info text-center">
            Loading...
          </div>
        :
          <blockquote className="blockquote text-right">
            <p className="mb-0">{quote}</p>
            <footer className="blockquote-footer">{author}</footer>
          </blockquote>
      }
      { !loading &&
        <div className="wraper-buttons">
          { counter > 1 &&
            <button
              className="btn btn-info"
              onClick={() => decrement()}
            >
              Back quote
            </button>
          }
          <button
            className="btn btn-warning"
            onClick={() => increment()}
          >
            Next quote
          </button>
        </div>
      }
    </>
  );
}

