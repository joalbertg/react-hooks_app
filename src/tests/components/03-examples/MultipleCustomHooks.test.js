import React from 'react';
import { shallow } from 'enzyme';
import { renderHook, act } from '@testing-library/react-hooks';

import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';

jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('Tests MultipleCustomHooks component', () => {
  useCounter.mockReturnValue({
    counter: 2
  });

  test('Should display correctly', () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null
    });

    const wrapper = shallow(<MultipleCustomHooks />);
    const divAlert = wrapper.find('.alert');

    expect(wrapper).toMatchSnapshot();
    expect(divAlert.exists()).toBe(true);
  });

  test('Should return info', () => {
    const data = [{
      author: 'Joalbert',
      quote: 'Hello World!!!'
    }];

    useFetch.mockReturnValue({
      data,
      loading: false,
      error: null
    });

    const wrapper = shallow(<MultipleCustomHooks />);
    const divAlert = wrapper.find('.alert');
    const quote = wrapper.find('blockquote p').text();
    const author = wrapper.find('blockquote footer').text();

    expect(wrapper).toMatchSnapshot();
    expect(divAlert.exists()).toBe(false);
    expect(quote).toBe(data[0].quote);
    expect(author).toBe(data[0].author);
  });

});

