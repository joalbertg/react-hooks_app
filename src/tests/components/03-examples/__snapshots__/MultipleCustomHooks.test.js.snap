// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests MultipleCustomHooks component Should display correctly 1`] = `
<Fragment>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
</Fragment>
`;

exports[`Tests MultipleCustomHooks component Should return info 1`] = `
<Fragment>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <blockquote
    className="blockquote text-right"
  >
    <p
      className="mb-0"
    >
      Hello World!!!
    </p>
    <footer
      className="blockquote-footer"
    >
      Joalbert
    </footer>
  </blockquote>
  <div
    className="wraper-buttons"
  >
    <button
      className="btn btn-info"
      onClick={[Function]}
    >
      Back quote
    </button>
    <button
      className="btn btn-warning"
      onClick={[Function]}
    >
      Next quote
    </button>
  </div>
</Fragment>
`;
